pkg/hostap/monitor.go

```diff
@@ -0,0 +1,239 @@
+				return
+			}
+			if size == 0 {
+				continue
+			}
+			eventStr := string(receivedByteArr[:size])
+			err = m.handleHostapdEvent(eventStr)
+			if err != nil {
+				level.Error(m.Logger).Log("hostapd-event", eventStr, "interface", m.IfName, "error", err)
+			}
+		}
+	}
+}
+
+func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
+	eventStrSlice := strings.Split(eventStr, " ")
+	if len(eventStrSlice) < 2 {
+		return nil
+	}
+	eventKeyStr := eventStrSlice[0][strings.Index(eventStrSlice[0], ">")+1:]
+	switch eventKeyStr {
+	case staConnectedEvent, eapSuccessEvent:
+		return m.handleAutheticatedMac(eventStrSlice[1])
+	case staDisconnectedEvent, eapFailureEvent:
+		return m.handleDeAutheticatedMac(eventStrSlice[1])
+	default:
+		level.Info(m.Logger).Log("hostapd-event", "unhandled event", m.IfName, eventStr)
+	}
+	return nil
+}
+
+func (m *InterfaceMonitor) handleAutheticatedMac(macAddress string) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	m.authenticatedMacs[macAddress] = nil
+	err := trafficcontrol.AllowTrafficFromMac(m.IfName, macAddress)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *InterfaceMonitor) handleDeAutheticatedMac(macAddress string) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	delete(m.authenticatedMacs, macAddress)
+	err := trafficcontrol.DenyTrafficFromMac(m.IfName, macAddress)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *InterfaceMonitor) sendKeepAlive() {
+	defer m.stopWg.Done()
+	for {
+		select {
+		case <-m.stop:
+			return
+		default:
+			time.Sleep(1 * time.Second)
+			m.hostApdConn.SetWriteDeadline(time.Now().Add(1 * time.Second))
+			_, err := m.hostApdConn.Write([]byte(pingCommand))
+			if err != nil {
+				if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+					continue
+				}
+				level.Error(m.Logger).Log("sockwrite", "error writing ping command to hostapd", m.IfName, err)
+				return
+			}
+		}
+	}
+}
+
+func (m *InterfaceMonitor) attachHostapd() error {
+	m.hostApdConn.SetWriteDeadline(time.Now().Add(1 * time.Second))
+	for {
+		_, err := m.hostApdConn.Write([]byte(attachCommand))
+		if err != nil {
+			if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+				continue
+			}
+			level.Error(m.Logger).Log("sockwrite", "error writing attach command to hostapd", m.IfName, err)
+			return err
+		}
+		break
+	}
+	return nil
+}
+
+func (m *InterfaceMonitor) handleIfEvents() {
+	defer m.stopWg.Done()
+	for {
+		select {
+		case <-m.stop:
+			close(m.ifEventCh)
+			return
+		case linkUpdateEvent := <-m.ifEventCh:
+			newOpState := linkUpdateEvent.Link.Attrs().OperState
+			if newOpState == m.operState {
+				level.Info(m.Logger).Log("interface", "event", m.IfName, "op state not changed", m.operState)
+				continue
+			}
+			m.operState = newOpState
+			level.Info(m.Logger).Log("interface", "event", m.IfName, "op state changed", m.operState)
+			m.mutex.Lock()
+			if newOpState == netlink.OperUp {
```

Comment:
I'm not sure if this logic is correct...

If a MAC is previously authenticated and the link goes down, we want to signal hostAPD that it's deauthenticated - If hostapd sends us a deauthenticated message, we can use handling of that message to block traffic.  If it doesn't, we'll have to additionally deny the traffic ourselves.  And we should flush `m.authenticatedMacs`.  When a link down happens, any previous MACs that had authenticated are no longer authenticated.

So that means that the bigger issue is what happens when the link comes back up... We do not want to immediately allow traffic from it again - We must wait for hostapd to do the actual 802.1x authentication and signal us that it's properly authenticated.  So we shouldn't ever call AlowTrafficFromMac in this handler, but only in the message handler for hostapd events.